rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/BaseLanguageServer.java

```diff
@@ -149,17 +152,21 @@ public void write(JsonWriter target, IValue value) throws IOException {
     }
 
     @SuppressWarnings({"java:S2189", "java:S106"})
-    public static void startLanguageServer(Supplier<IBaseTextDocumentService> service, int portNumber) {
+    public static void startLanguageServer(ExecutorService threadPool, Function<ExecutorService, IBaseTextDocumentService> docServiceProvider, BiFunction<ExecutorService, IBaseTextDocumentService, BaseWorkspaceService> workspaceServiceProvider, int portNumber) {
         logger.info("Starting Rascal Language Server: {}", getVersion());
 
+        var docService = docServiceProvider.apply(threadPool);
+        var wsService = workspaceServiceProvider.apply(threadPool, docService);
+        docService.pair(wsService);
+
         if (DEPLOY_MODE) {
-            startLSP(constructLSPClient(capturedIn, capturedOut, new ActualLanguageServer(() -> System.exit(0), service.get())));
+            startLSP(constructLSPClient(capturedIn, capturedOut, new ActualLanguageServer(() -> System.exit(0), docService, wsService)));
         }
         else {
             try (ServerSocket serverSocket = new ServerSocket(portNumber, 0, InetAddress.getByName("127.0.0.1"))) {
                 logger.info("Rascal LSP server listens on port number: {}", portNumber);
                 while (true) {
-                    startLSP(constructLSPClient(serverSocket.accept(), new ActualLanguageServer(() -> {}, service.get())));
+                    startLSP(constructLSPClient(serverSocket.accept(), new ActualLanguageServer(() -> {}, docService, wsService)));
```

Comment:
Since this happens in a loop, the semantics have now changed.

This code exists so when developing 2nd level, you can restart VS Code, and then we automatically restart the LSP instance to get a fresh one.  So we should reconstruct them every time we start the loop.

Reply:
Right. Fixed!